### PR TITLE
Split tests for webpack 5

### PR DESCRIPTION
**Pull request by smelukov**
This PR contains a:

- [ ] **bugfix**
- [ ] new **feature**
- [x] **code refactor**
- [x] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### CODE REVIEW

File: test/sourceMap-option.test.js [hunk 1]
```diff
@@ -201,20 +205,22 @@ describe('sourceMap', () => {
     expect(getWarnings(stats)).toMatchSnapshot('warnings');
   });
 
-  it('should match snapshot when the "devtool" option has the "sourcemap" value', async () => {
-    const compiler = getCompiler({
-      entry: `${__dirname}/fixtures/entry.js`,
-      devtool: 'sourcemap',
-    });
+  if (getCompiler.isWebpack4()) {
```
Comment (evilebottnawi) [inline test/sourceMap-option.test.js]: Why only webpack@4?

File: test/sourceMap-option.test.js [hunk 2]
```diff
@@ -137,54 +137,58 @@ describe('sourceMap', () => {
...
-          }
-        );
-      }
-    })();
-    const compiler = getCompiler({
-      entry: `${__dirname}/fixtures/entry.js`,
-      devtool: 'source-map',
-      plugins: [emitBrokenSourceMapPlugin],
-    });
+  if (getCompiler.isWebpack4()) {
```
Comment (evilebottnawi) [inline test/sourceMap-option.test.js]: Webpack@5 already throw error on broken source maps, right?

File: test/parallel-option-failure.test.js [hunk 3]
```diff
@@ -76,21 +80,23 @@ describe('parallel option', () => {
     expect(getWarnings(stats)).toMatchSnapshot('warnings');
   });
 
-  it('should match snapshot for errors into the "jest-worker" package whe the "cache" option is "true"', async () => {
-    new TerserPlugin({ parallel: true, cache: true }).apply(compiler);
+  if (getCompiler.isWebpack4()) {
```
Comment (evilebottnawi) [inline test/parallel-option-failure.test.js]: Why webpack@4 only?

File: test/sourceMap-option.test.js (hunk 1 above)
Comment (smelukov) [reply test/sourceMap-option.test.js]: Cause devtool: sourcemap is not valid value anymore

File: test/helpers/snapshotResolver.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
+const path = require('path');
+
+const webpackPackageJson = require('webpack/package.json');
+
+// eslint-disable-next-line global-require
+const [webpackVersion] = webpackPackageJson.version;
+const snapshotExtension = `.snap.webpack${webpackVersion}`;
```
Comment (evilebottnawi) [inline test/helpers/snapshotResolver.js]: Let's use `.webpack4.snap`, some IDE highlight snap based `snap` extension

File: test/sourceMap-option.test.js (hunk 2 above)
Comment (smelukov) [reply test/sourceMap-option.test.js]: source-map package does it

File: test/helpers/getCompiler.js [hunk 5]
```diff
@@ -34,3 +35,5 @@ export default function getCompiler(options = {}) {
 
   return compiler;
 }
+
+getCompiler.isWebpack4 = () => webpackPackageJson.version[0] === '4';
```
Comment (evilebottnawi) [inline test/helpers/getCompiler.js]: `https://github.com/webpack/webpack/blob/v4.41.2/lib/webpack.js#L80

File: test/helpers/snapshotResolver.js (hunk 4 above)
Comment (smelukov) [reply test/helpers/snapshotResolver.js]: It does not work on CI, cause jest tries to absolete webpack 5 snapshot (jest finds these by snap-extension) when runs webpack 4 test and throws an error

File: test/cache-option.test.js [hunk 6]
```diff
@@ -27,303 +27,307 @@ const otherOtherOtherCacheDir = findCacheDir({
...
-    compiler = getCompiler({
-      entry: {
-        one: `${__dirname}/fixtures/cache.js`,
-        two: `${__dirname}/fixtures/cache-1.js`,
-        three: `${__dirname}/fixtures/cache-2.js`,
-        four: `${__dirname}/fixtures/cache-3.js`,
-        five: `${__dirname}/fixtures/cache-4.js`,
-      },
-    });
+if (getCompiler.isWebpack4()) {
```
Comment (evilebottnawi) [inline test/cache-option.test.js]: Let's keep testing `cache` here for webpack@4 and webpack@5, and fix cache tests in PR where we implement new cache API for webpack@5

File: test/__snapshots__/warningsFilter-option.test.js.snap.webpack5 [hunk 7]
```diff
@@ -0,0 +1,77 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`warningsFilter option should match snapshot for a "function" value and the "sourceMap" option is "false" (filter by message): assets 1`] = `
+Object {
+  "one.js": "",
+  "two.js": "",
```
Comment (evilebottnawi) [inline test/__snapshots__/warningsFilter-option.test.js.snap.webpack5]: Empty files? Bug?

File: test/TerserPlugin.test.js [hunk 8]
```diff
@@ -185,52 +186,93 @@ describe('TerserPlugin', () => {
...
-  it('should regenerate hash', async () => {
-    const originalMainTemplateUpdateHashForChunk =
-      MainTemplate.prototype.updateHashForChunk;
-    const originalChunkTemplateUpdateHashForChunk =
-      ChunkTemplate.prototype.updateHashForChunk;
-    const mockMainTemplateUpdateHashForChunk = jest.fn();
-    const mockChunkTemplateUpdateHashFocChunk = jest.fn();
+  if (getCompiler.isWebpack4()) {
+    it('should regenerate hash', async () => {
+      const originalMainTemplateUpdateHashForChunk =
```
Comment (evilebottnawi) [inline test/TerserPlugin.test.js]: Put `getCompiler.isWebpack4())` inside `it`, if/else tests is bad practice

File: azure-pipelines.yml [hunk 9]
```diff
@@ -199,7 +199,7 @@ jobs:
         displayName: 'Install dependencies'
       - script: npm i webpack@$(webpack_version)
         displayName: 'Install "webpack@$(webpack_version)"'
-      - script: npm run test:coverage -- --ci --reporters="default" --reporters="jest-junit" || $(continue_on_error)
+      - script: npm run test:coverage -- --ci --snapshot-resolver="./test/helpers/snapshotResolver.js" --reporters="default" --reporters="jest-junit" || $(continue_on_error)
```
Comment (evilebottnawi) [inline azure-pipelines.yml]: Why we need `--snapshot-resolver` here?

File: test/__snapshots__/warningsFilter-option.test.js.snap.webpack5 (hunk 7 above)
Comment (smelukov) [reply test/__snapshots__/warningsFilter-option.test.js.snap.webpack5]: Or webpack 5 just emit more efficient es6-code that drops by terser

File: azure-pipelines.yml (hunk 9 above)
Comment (smelukov) [reply azure-pipelines.yml]: It does not work without it when ci flag is used

File: test/sourceMap-option.test.js (hunk 2 above)
Comment (evilebottnawi) [reply test/sourceMap-option.test.js]: :+1:

File: test/sourceMap-option.test.js (hunk 1 above)
Comment (evilebottnawi) [reply test/sourceMap-option.test.js]: Let's fix it in test - rename `sourcemap` to  `source-map`

File: test/__snapshots__/warningsFilter-option.test.js.snap.webpack5 (hunk 7 above)
Comment (evilebottnawi) [reply test/__snapshots__/warningsFilter-option.test.js.snap.webpack5]: Great, I thought so

File: azure-pipelines.yml (hunk 9 above)
Comment (evilebottnawi) [reply azure-pipelines.yml]: :+1:

**Pull request merged**